feat: add generic iframe

DanielHabenicht opened the pull request:
This will add a generic IFrame quite similar to the video functionality, but with the following in mind:
- the user can't configure the iframe string, it is set similar to the etherpad feature
- you can use pre defined template strings for `roomname` and `language` (which can be extended in the future)
- makes another application available in jitsi (because etherpad and the whiteboard are quite common requirements)

This will fix #5295. 

But I would suspect people will overwrite it in their deployment to fit their application name (e.g. `Open shared whiteboard`)


see #9286 for reference and former discussions

Can be tested by adding the following Config: 
```
// Add in app.js or react/features/app/actions.native.js:129
// eslint-disable-next-line no-undef
config = {
    // eslint-disable-next-line no-undef
    ...config,
    sharedIFrames: {
        restrictControlToInitialSharer: false,
        frames: {
            whiteboard: { // The Id of the sharedIframe
                title: 'Whiteboard',
                templateUrl: 'https://wbo.ophir.dev/boards/{room}?lang={lang}',
                avatarUrl: 'images/sharedIFrameAvatar.svg'
            },
            etherpad: {
                title: 'Document',
                templateUrl: 'https://yopad.eu/p/{room}',
                avatarUrl: 'images/sharedIFrameAvatarEtherpad.svg'
            }
       }
    },
};
```

jitsi-jenkins commented:
Hi, thanks for your contribution!
If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)? We would unfortunately be unable to merge your patch unless we have that piece :(.

DanielHabenicht commented:
Open Questions: 
- [ ] Should the etherpad implementation also be moved into this generic implementation (e.g. making this configurable for more than one IFrame?)

damencho commented:
> Open Questions:
> 
> * [ ]  Should the etherpad implementation also be moved into this generic implementation (e.g. making this configurable for more than one IFrame?)

The etherpad has some more in it, like some mouse move handling I remember ... but if they can share code without breaking people's deployments that is the ideal situation.

damencho commented:
Ok, I will try to take a look today.

damencho commented:
Can you give me a simple example that I can use for testing? Thanks.

damencho commented:
You need to correctly sort main.json changes.

DanielHabenicht commented:
> Can you give me a simple example that I can use for testing? Thanks.

You're referring to an environment where you can test, with this feature configured?
If so, can you send me your email at daniel.habenicht@t-systems.com so I can send you credentials for accessing it?

DanielHabenicht commented:
> You need to correctly sort main.json changes.

done!

damencho commented:
There are still lint errors. You can check them locally with `npm run lint`.

https://github.com/jitsi/jitsi-meet/blob/39793110b68da40dd82626b9d4f5f568fb156d5b/.github/workflows/ci.yml#L23

DanielHabenicht commented:
> There are still lint errors. You can check them locally with `npm run lint`.
> 
> https://github.com/jitsi/jitsi-meet/blob/39793110b68da40dd82626b9d4f5f568fb156d5b/.github/workflows/ci.yml#L23

will do

Also user tests we came across the following questions: 

1. If the IFrame is already shared and a new user joins, the IFrame is automatically shown instead of the tile view. 
    I inherited this behaviour from the shared-video but most of the users told us that it is inconvenient. So I would change it back to the normal behaviour as it is with normal users. (One could think about doing the same for the shared-video) 
    But still retaining the functionality of focusing the shared-iframe when it is first shared. 
    Do you agree with that?
2. I also inherited the behaviour for disabling the button if another user has already shared the iframe. Users also found this confusing. There are three options: 
     - (Current) The button is greyed out. An aria label is displayed explaining why it is greyed out "Someone else is already sharing."
     - (A) Scrape the whole rights logic and let everyone shared or unshare the iframe
     - (B) Hide the button instead of greying it out
  Whats your opinion?

damencho commented:
> 1. If the IFrame is already shared and a new user joins, the IFrame is automatically shown instead of the tile view.
>    I inherited this behaviour from the shared-video but most of the users told us that it is inconvenient. So I would change it back to the normal behaviour as it is with normal users. (One could think about doing the same for the shared-video)
>    But still retaining the functionality of focusing the shared-iframe when it is first shared.
>    Do you agree with that?

Sounds good. I hope it does not complicate the logic outside the feature code.





> I also inherited the behaviour for disabling the button if another user has already shared the iframe. Users also found this confusing. There are three options:
> * (Current) The button is greyed out. An aria label is displayed explaining why it is greyed out "Someone else is already sharing."
> * (A) Scrape the whole rights logic and let everyone shared or unshare the iframe
> * (B) Hide the button instead of greying it out
>   Whats your opinion?

I would say keep it the same as the shared video.

DanielHabenicht commented:
> > 1. If the IFrame is already shared and a new user joins, the IFrame is automatically shown instead of the tile view.
> >    I inherited this behaviour from the shared-video but most of the users told us that it is inconvenient. So I would change it back to the normal behaviour as it is with normal users. (One could think about doing the same for the shared-video)
> >    But still retaining the functionality of focusing the shared-iframe when it is first shared.
> >    Do you agree with that?
> 
> Sounds good. I hope it does not complicate the logic outside the feature code.

At least one more if statement ^

> > I also inherited the behaviour for disabling the button if another user has already shared the iframe. Users also found this confusing. There are three options:
> > 
> > * (Current) The button is greyed out. An aria label is displayed explaining why it is greyed out "Someone else is already sharing."
> > * (A) Scrape the whole rights logic and let everyone shared or unshare the iframe
> > * (B) Hide the button instead of greying it out
> >   Whats your opinion?
> 
> I would say keep it the same as the shared video.

Alright so the current option (but I already added the label, which is the only difference from shared-video)

damencho commented:
Have you tried it on mobile? I see the participant and its name but nothing is being displayed...

damencho commented on react/features/base/participants/components/ParticipantView.native.js:
```diff
@@ -208,11 +220,17 @@ class ParticipantView extends Component<Props> {
+        const renderSharedVideo = _isFakeParticipant
+            && (
+                _participantName === VIDEO_PLAYER_PARTICIPANT_NAME
+                || _participantName === YOUTUBE_PLAYER_PARTICIPANT_NAME)
```
Why did you need to change the logic for `renderSharedVideo `?

damencho commented on react/features/base/participants/components/ParticipantView.native.js:
```diff
@@ -208,11 +220,17 @@ class ParticipantView extends Component<Props> {
+                || _participantName === YOUTUBE_PLAYER_PARTICIPANT_NAME)
+            && !disableVideo;
+        const renderSharedIFrame = _isFakeParticipant
+            && _participantName === _iFrameParticipantName && !disableVideo;
```
Yeah, why do we need the participant name check?

damencho commented on react/features/large-video/components/LargeVideo.web.js:
```diff
@@ -126,31 +151,50 @@ class LargeVideo extends Component<Props> {
+                || _participantName === YOUTUBE_PLAYER_PARTICIPANT_NAME);
+        const renderSharedIFrame = _isFakeParticipant
+            && _participantName === _iFrameParticipantName;
+
```
Same here.

damencho replied:
A here is to be able to distinguish and set the style ... I see.

damencho commented on config.js:
```diff
@@ -1163,6 +1165,16 @@ var config = {
+    //  - {lang} - The ISO Language tag
+    //
+    // sharedIFrameTemplateUrl: 'https://example.com/link/to/your/iframe/app/{room}?lang={lang}',
+    // sharedIFrameName: 'Whiteboard',
```
Can you use sharedIFrameName for the button name, so it shows 'Start Whiteboard'?

DanielHabenicht replied:
Sure, i will modify the buttons.

DanielHabenicht replied:
This is now configurable in the main language files, based on the key used for the iframe, see: 
https://github.com/jitsi/jitsi-meet/pull/11052/commits/0d41f018ec0182f621772ab2d72efc27c218db4f#diff-acc6f5186c27e0f1c972135d39d2567f7deea1a3993c13eb8df685a955caa869

damencho replied:
Do you see a problem with dynamically using the name coming form the config?
You cannot use the language files without rebuilding. And this is being configured from config.js ... If my service is PacMan in iframe, I cannot set a name without rebuilding the sources ...

DanielHabenicht replied:
Ugh, I thought it would be at runtime. I did not deploy the update to our server yet.

damencho replied:
Nope :) the main.json is embed in the app.bundle for web. On mobile all of the language files are embedded in the app.

DanielHabenicht replied:
Ahh, this clears up another problem I had. Now I know why it does not update the english version of my app. 
But why is it used as english then (instead of as a fallback, when no other string is supplied)? 
So it would work in every language except English...
Should I go back to the stitching approach of using the participant name? and a prefix/suffix button string combination. 
I thought it would make more sense to supply the user with the actual word in their language instead of with the same in every language.

damencho replied:
> I thought it would make more sense to supply the user with the actual word in their language instead of with the same in every language.

There is no mechanism to add this dynamically at the moment.



> Should I go back to the stitching approach of using the participant name? (a prefix/suffix button string combination).

If you mean to generate the button name from the config string - yes.



> But why is it used as english then (instead of as a fallback, when no other string is supplied)?

I didn't get that.

DanielHabenicht replied:
> There is no mechanism to add this dynamically at the moment.

For me it is already working. I tried again on our environment and all updates to the json files get applied without recompiling, except the one from the main.json. 

> If you mean to generate the button name from the config string - yes.

I will update as you suggested. 

>  > But why is it used as english then (instead of as a fallback, when no other string is supplied)?
>
> I didn't get that.

The main-en.json does not seem to get used. It always uses the english words from main.json. 
I thought the main.json strings are used as fallback, if a language does not specify a translation string.

damencho commented:
What about mobile? Were you testing it there?

saghul commented:
> * Should the etherpad implementation also be moved into this generic implementation (e.g. making this configurable for more than one IFrame?)
>       ==> Doing this in a different PR

So, I'd like to see this done now, otherwise we might miss requirements here. As in, how do we know we built the right abstraction, if there is no single implementation of it?

DanielHabenicht commented:
> > * Should the etherpad implementation also be moved into this generic implementation (e.g. making this configurable for more than one IFrame?)
> >   ==> Doing this in a different PR
> 
> So, I'd like to see this done now, otherwise we might miss requirements here. As in, how do we know we built the right abstraction, if there is no single implementation of it?

My way of thinking was keeping this PR bound to one change. And making another for the removal of the etherpad option (migrating it to this). This would bring the benefit of making this a non-breaking change and not changing to many things at the same time. I already tried the generic implementation with a linked Etherpad, but will have a look at making multiple applications available at the same time.

saghul reviewed (commented):
I like this a lot, but I have 3 concerns:

- Web currently models Etherpad and YT sharing in a similar way: a fake participant, except that in case of Etherpad there is no participant thumbnail. On mobile, YT sharing is a fake participant, but Etherpad is a screen which covers the entire conference. If interaction is necessary (as it is), we should probably implement this the same way as Etherpad.

- Having a generic option which allows a single use sounds like an odd limitation. I think of this as "shared apps", so we should probably support specifying multiple URls, each with a tiltle

- It doesn't seem to be fully implemented on mobile

DanielHabenicht commented on config.js:
```diff
@@ -1163,6 +1165,16 @@ var config = {
     // etherpad_base: 'https://your-etherpad-installati.on/p/',
 
+
+    // Generic IFrame to be used for any application you like. You can name the menu entries in "/lang/main.json"
```
yes!

DanielHabenicht replied:
Fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/0d41f018ec0182f621772ab2d72efc27c218db4f

DanielHabenicht commented:
> What about mobile? Were you testing it there?

You mean the mobile app or Mobile View? Is there a nice way of developing it without android studio?

saghul commented:
> > What about mobile? Were you testing it there?
> 
> You mean the mobile app or Mobile View? Is there a nice way of developing it without android studio?

I mean the mobile apps. They are written in JS as part of this repo (the .native.js files) you can use Xcode or Android Studio to run them, but you can develop with any editor really since the bulk of the code is JS. You can check the current Etherpad implementation as a starter.

DanielHabenicht commented on modules/UI/videolayout/VideoContainer.js:
```diff
@@ -389,9 +389,11 @@ export class VideoContainer extends LargeContainer {
-            $element.css(
-                'top',
-                $avatarImage.offset().top + $avatarImage.height() + 10);
+            if ($avatarImage.length > 0) {
```
It somehow popped up while trying it out: https://github.com/DanielHabenicht/jitsi-meet/commit/ad69885e3db3e2e02ada6ceda5d4ab0ef1137df2
But I can't reproduce anymore.

DanielHabenicht replied:
Should I remove it?

damencho commented:
> * Web currently models Etherpad and YT sharing in a similar way: a fake participant, except that in case of Etherpad there is no participant thumbnail. On mobile, YT sharing is a fake participant, but Etherpad is a screen which covers the entire conference. If interaction is necessary (as it is), we should probably implement this the same way as Etherpad.
> *

@saghul I asked explicitly for this. As Etherpad is buggy and we break it a lot as we do not use it. And the fake participant we use for youtube sharing and we will be using something similar (I think) for multistreaming ...

DanielHabenicht commented on react/features/large-video/components/LargeVideo.web.js:
```diff
@@ -126,31 +151,50 @@ class LargeVideo extends Component<Props> {
                 <div id = 'etherpad' />
 
-                <Watermarks />
+                { !renderSharedIFrame && <Watermarks /> }
```
Well all other implementations simply used a higher `z-index` to draw over the logo: 
https://github.com/jitsi/jitsi-meet/blob/7e5ffdb3906d80bdbf2976d5f85d3e79f7e6455a/modules/UI/etherpad/Etherpad.js#L146
But the baseline is, for a generic Implementation it's not guaranteed that the application will have a background at all. Thus simply having a higher z-index will not help.

DanielHabenicht replied:
That's the base problem without removing it: 
![image](https://user-images.githubusercontent.com/13590797/160412765-86accd7b-e0c0-4e28-8ac3-7fc6f915257b.png)

damencho commented:
> So is there a way to run it just with node? something like `make dev --mobile`?

I'm not a mobile developer! But what I do when I need it - I run it on my personal Android phone.
Make sure you have java 11 installed and JAVA_HOME set and no Jitsi Meet app installed on the phone. 
And then I do: `npx react-native run-android` (of course USB debugging turned on).

DanielHabenicht commented on react/features/shared-iframe/functions.js:
```diff
@@ -0,0 +1,46 @@
+    const state = toState(stateful);
+    const { sharedIFrameName } = state['features/base/config'];
+
+    // eslint-disable-next-line no-unused-vars
```
How do you mean? It is destructuring and naming the `Map<string, Participant>` return value of the function. I don't see how Object.values would help. So I am interested in what you meant! 
```js
for (const participant of getFakeParticipants(stateful)) {
        if (participant[1].name === sharedIFrameName || participant[1].name === IFRAME_PLAYER_PARTICIPANT_NAME) {
            isIFrameActive = true;
            break;
        }
    }
```

saghul replied:
Ah I thought it was an object! In that case you can do `theMap.values()` and you'll get an iterator over the values, which is what you need.

DanielHabenicht replied:
I tried it with:
```
for (const { p } of getFakeParticipants(stateful).values()) {
    if (p.name === sharedIFrameName || p.name === IFRAME_PLAYER_PARTICIPANT_NAME) {
        isIFrameActive = true;
        break;
    }
}
```
But this will throw an error property not found on `undefined`.

DanielHabenicht commented on react/features/shared-iframe/components/IFrameManager.js:
```diff
@@ -0,0 +1,149 @@
+ * @param {Object} state - The Redux state.
+ * @returns {Props}
+ */
+function _mapStateToProps(): $Shape<Props> {
```
fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

DanielHabenicht commented on react/features/shared-iframe/components/IFrameManager.js:
```diff
@@ -0,0 +1,149 @@
+/* @flow */
```
Interestingly, the linter fails if I remove it: 
```bash
/workspaces/jitsi-meet/react/features/shared-iframe/components/IFrameManager.js
  129:51  error  '$Shape' is not defined  no-undef
```
But this should be taken care of after using the right `dispatch` pattern.

DanielHabenicht replied:
fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

DanielHabenicht commented on react/features/shared-iframe/components/IFrameManager.js:
```diff
@@ -0,0 +1,149 @@
+/* @flow */
+/* eslint-disable no-invalid-this */
```
removed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

DanielHabenicht commented on react/features/shared-iframe/components/IFrameManager.js:
```diff
@@ -0,0 +1,149 @@
+ * @param {Function} dispatch - The Redux dispatch function.
+ * @returns {Props}
+ */
+function _mapDispatchToProps(dispatch: Function): $Shape<Props> {
```
Mhm, this does not work. Do I need to populate dispatch somehow? It's not defined if I just define and use it.

DanielHabenicht commented on react/features/shared-iframe/reducer.native.js:
```diff
@@ -0,0 +1,28 @@
+
+import { RESET_SHARED_IFRAME_STATUS, SET_SHARED_IFRAME_STATUS } from './actionTypes';
+
+const initialState = {};
```
fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

DanielHabenicht commented on react/features/shared-iframe/middleware.any.js:
```diff
@@ -0,0 +1,167 @@
+    const localParticipantId = getLocalParticipant(getState()).id;
+
+    if (isSharing === 'true') {
+
```
fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

DanielHabenicht commented on react/features/shared-iframe/components/SharedIFrame.js:
```diff
@@ -0,0 +1,177 @@
+ *
+ * @augments Component
+ */
+class SharedIFrame extends Component<Props> {
```
fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

DanielHabenicht commented on react/features/base/participants/components/ParticipantView.native.js:
```diff
@@ -277,6 +297,7 @@ function _mapStateToProps(state, ownProps) {
                 || JitsiParticipantConnectionStatus.ACTIVE,
         _isFakeParticipant: participant && participant.isFakeParticipant,
         _participantName: participantName,
+        _iFrameParticipantName: sharedIFrameName || IFRAME_PLAYER_PARTICIPANT_NAME,
```
fixed in https://github.com/jitsi/jitsi-meet/pull/11052/commits/46c78fcd20f9d6d7ff6084a644f9d0e8ddc42b61

damencho commented on react/features/shared-iframe/middleware.any.js:
```diff
@@ -0,0 +1,167 @@
+            name: sharedIFrameName || IFRAME_PLAYER_PARTICIPANT_NAME
+        }));
+
+        dispatch(pinParticipant(iFrameTemplateUrl));
```
If you receive CONFERENCE_JOINED and there was an iframe start command before that - this is when you are joining in a meeting when someone started the iframe sharing. If you receive the iframe command after you had received CONFERENCE_JOINED this means someone is starting the iframe thingy during the meeting.

DanielHabenicht replied:
implemented in https://github.com/jitsi/jitsi-meet/pull/11052/commits/5fb3f9479a948130c6feb4021a993a579e46d18c

DanielHabenicht commented:
> > * Should the etherpad implementation also be moved into this generic implementation (e.g. making this configurable for more than one IFrame?)
> >   ==> Doing this in a different PR
> 
> So, I'd like to see this done now, otherwise we might miss requirements here. As in, how do we know we built the right abstraction, if there is no single implementation of it?

@saghul I added a commit for this. Its testable but there is still one bug concerning the switching of fake participants - it won't change between them, although I set it to be the only focused user: 
https://github.com/jitsi/jitsi-meet/pull/11052/commits/0d41f018ec0182f621772ab2d72efc27c218db4f#diff-915778d31824bfed939a12755b3e97ae0addefa8acfb6c45127ae8e914ef55c4

Any idea what I am doing wrong? 

Regarding the next steps: 
1. Please confirm that I can go ahead with this implementation
2. I will add and test the mobile view (finally got it to work on my non windows PC) 
3. Fix some Bugs 
   - disable focus of iframes for newly joined users
   - test whiteboard movement not possible in iframe
   - test again on mobile touch devices

DanielHabenicht commented:
I think I got the error fixed and it should work now (except in the native app - as this is not yet implemented). (As I am hitting a time limit from my company, would you be open to merging it without native support?) 

Please have a look at the already answered comments and write back to me, I hope I answered them all.

DanielHabenicht commented:
@damencho I updated it, for me the branch is working without any issues. Can you share the error?

damencho commented:
Checking again :)

damencho commented:
```
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at Function.mapToProps (LargeVideo.web.js:362:38)
    at a (wrapMapToProps.js:41:92)
    at Function.a.mapToProps (wrapMapToProps.js:50:19)
    at a (wrapMapToProps.js:41:46)
    at selectorFactory.js:22:18
    at connectAdvanced.js:210:16
    at Object.useMemo (react-dom.production.min.js:172:115)
    at t.useMemo (react.production.min.js:23:113)
    at d (connectAdvanced.js:195:30)
    at io (react-dom.production.min.js:157:137)
    at Ho (react-dom.production.min.js:180:154)
    at qo (react-dom.production.min.js:178:169)
    at Fo (react-dom.production.min.js:177:178)
    at Ws (react-dom.production.min.js:274:126)
    at Al (react-dom.production.min.js:250:347)
    at xl (react-dom.production.min.js:250:278)
    at El (react-dom.production.min.js:250:138)
    at gl (react-dom.production.min.js:243:163)
    at react-dom.production.min.js:123:115
    at t.unstable_runWithPriority (scheduler.production.min.js:18:343)
    at Wr (react-dom.production.min.js:122:325)
    at Kr (react-dom.production.min.js:123:61)
    at Gr (react-dom.production.min.js:122:428)
    at dl (react-dom.production.min.js:237:203)
    at Object.enqueueSetState (react-dom.production.min.js:133:211)
    at APP.g.setState (react.production.min.js:12:369)
    at BaseApp.js:263:18
    at new Promise (<anonymous>)
    at APP._navigate (BaseApp.js:262:16)
    at middleware.js:154:48
ss @ react-dom.production.min.js:216
07:58:25.862 
```

damencho commented:
There are also a few lint errors.

damencho commented:
sharedIFrameConfig is undefined.

damencho commented:
A you changed the way it is configured ... but anyway this error should be fixed as those not configuring it will break their meetings

damencho commented on react/features/shared-iframe/components/SharedIFrame.js:
```diff
@@ -0,0 +1,181 @@
+        let width;
+        let height;
+
+        if (interfaceConfig.VERTICAL_FILMSTRIP) {
```
This breaks mobile.

DanielHabenicht replied:
what exactly is breaking? 
I have no problems opening it in chrome emulated as pixel 5

DanielHabenicht replied:
![image](https://user-images.githubusercontent.com/13590797/166216462-c1085924-0a12-482f-a75a-7c14312d6484.png)

saghul replied:
The mobile app, not the mobile web view.

damencho commented on react/features/shared-iframe/components/SharedIFrame.js:
```diff
@@ -0,0 +1,181 @@
+            height = `${clientHeight - getToolboxHeight()}px`;
+        } else {
+            if (filmstripVisible) {
+                height = `${clientHeight - Filmstrip.getFilmstripHeight()}px`;
```
This call also breakes mobile.

DanielHabenicht commented:
> convert undefined or null to object

fixed it
also fixed the translation issue

Will try to take a look at the app again, but as I said earlier I have no time left to work on it.

matbgn commented:
Such an amazing job done by @DanielHabenicht, simply THANK YOU!

It will be a shame if we could not merge all this job done and let move away master branch. What are your remaining expectations @damencho & @saghul?

DanielHabenicht commented:
@saghul @damencho @farhan711 I implemented the feature also for the native app. 
Also I had to squash the commits to apply them to an earlier version of jitsi. 
I fixed the lint and lang-sort issues that came from merging all of it together. 

With this the feature should really be ready to go. 
Please note that I have only on month left to merge this, afterwards I will be on sabbatical for 4 months.

saghul commented:
Thanks a lot @DanielHabenicht ! We have started exploring integrating a whiteboard app into Jitsi Meet so we'll be looking at this PR as the foundation really soon. You can do the languages on a separate PR once this one lands. Thanks for the work so far and the heads up on the timeline, I think we can make it, but if we don't, we'll wrap it up, no problem!

@mihhu this is the PR we talked about.

matbgn commented:
@saghul Do I understand correctly that you do not want to introduce a generic solution but only dedicated to the whiteboard?

saghul commented:
> @saghul Do I understand correctly that you do not want to introduce a generic solution but only dedicated to the whiteboard?

On the contrary, we want _this_ generic solution, and then use it for integrating a white-boarding solution.

mihhu reviewed (commented):
Thank you for your contribution! I've done some testing on the web and left some ideas down below. Can you please tell me how you tested on the native app? On my side, the iframe is not displayed, only the avatar with the initial of the key.

There are also two more things that should be covered later on:
- the more menu should be dismissed
- participants that join after the iframe being shared should also see the iframe in stage view

DanielHabenicht commented:
> Thank you for your contribution! I've done some testing on the web and left some ideas down below. Can you please tell me how you tested on the native app? On my side, the iframe is not displayed, only the avatar with the initial of the key and the Shared document screen doesn't render anything.

Did you click on the avatars? It should open the iframe in a new pane, otherwise there would be to much cluttered around the view.

> There are also two more things that should be covered later on:
> 
> * the more menu should be dismissed

I will have a look!

> * participants that join after the iframe being shared should also see the iframe in stage view

That should already be the case, I verify that its working.

mihhu commented:
> Did you click on the avatars? It should open the iframe in a new pane, otherwise there would be to much cluttered around the view.

I agree, opening the iframe in a new pane is ideal in mobile conditions (and implies having the moderator opening the pane for the newly joined participants as they show up). I think I'm missing something, because I don't see anything in the new pane:

![image (3)](https://user-images.githubusercontent.com/20574669/177542820-f33ac221-2237-4517-a6a4-30b6eb81ed81.png)

Thank you!

DanielHabenicht commented on config.js:
```diff
@@ -1252,6 +1254,23 @@ var config = {
+    //  - {room} - The room id
+    //  - {lang} - The ISO Language tag
+    //
+    // sharedIFrameConfig: {
```
done in 0c28a421ea580e7f3858ca4bb7c2bfb8adeacefc and 8517487d254a4f954cd0f3c42dfae7bc9b32db10

DanielHabenicht commented on config.js:
```diff
@@ -1252,6 +1254,23 @@ var config = {
+    //  - {lang} - The ISO Language tag
+    //
+    // sharedIFrameConfig: {
+    //     whiteboard: { // The Id and displayed name of the sharedIframe
```
![image](https://user-images.githubusercontent.com/13590797/177948490-69f8247e-1a9d-465b-9d8d-412878793e8e.png)
![image](https://user-images.githubusercontent.com/13590797/177948533-c3c2f2ef-2e9c-475f-9dab-0bf93d2299c8.png)
I updated it for all `fakeParticipants` which includes the current Video and Etherpad implementation. 
in 80d92808623944569320568ba8ca378ddb3b30cd and b165f4d3bd4f8d32980359477bae4c7dbb2ae7c1

mihhu replied:
Thank you for updating the Video part as well! I'm not entirely sure about the messages though. If you could make this configurable too, it would be great, but it's ok if you leave it at that.

DanielHabenicht replied:
it configurable in the `main.json` via the `notify.shareStart|shareStopped` keys. 
If you have a suggestion on better wording, please name it. I am happy to edit it. I am not that good at wording it clearly

mihhu replied:
Ah awesome! Haha me neither, we'll see with the UX for that. Thanks!

DanielHabenicht commented on modules/API/API.js:
```diff
@@ -463,6 +464,18 @@ function initCommands() {
             APP.store.dispatch(stopSharedVideo());
         },
 
+        'toggle-share-iframe': () => {
```
320ba10233797c7b2610f1d3e9f8a28196618d5d

DanielHabenicht commented on modules/API/API.js:
```diff
@@ -463,6 +464,18 @@ function initCommands() {
+        },
+
+        'stop-share-iframe': shareKey => {
+            logger.debug('Share iframe command received');
```
320ba10233797c7b2610f1d3e9f8a28196618d5d

DanielHabenicht commented on react/features/base/flags/constants.js:
```diff
@@ -80,6 +80,12 @@ export const FULLSCREEN_ENABLED = 'fullscreen.enabled';
+ * Flag indicating if the video share button should be enabled
+ * Default: enabled (true).
+ */
+export const IFRAME_SHARE_BUTTON_ENABLED = 'iframe-share.enabled';
```
[b7ffcae6f17cbe1bb29629f12d0633aba1b5c296](https://github.com/jitsi/jitsi-meet/commit/b7ffcae6f17cbe1bb29629f12d0633aba1b5c296)

DanielHabenicht commented:
> > There are also two more things that should be covered later on:
> > 
> > * the more menu should be dismissed
> 
> I will have a look!

@mihhu Can you help me out here? I don't get why the `afterClick` event from the [OverflowMenu](https://github.com/DanielHabenicht/jitsi-meet/blob/21cf7f23c2b71eab5c36ff2f6c8c9a33fbd10953/react/features/toolbox/components/native/OverflowMenu.js#L113) is not passed down to the [SharedIFrameButtonContainer](https://github.com/DanielHabenicht/jitsi-meet/blob/b7ffcae6f17cbe1bb29629f12d0633aba1b5c296/react/features/shared-iframe/components/SharedIFrameButtonContainer.js#L31) `props`, do you happen to know why?

DanielHabenicht commented on react/features/shared-iframe/actions.any.js:
```diff
@@ -0,0 +1,160 @@
+            dispatch(setSharedIFrameStatus({
+                shareKey,
+                iFrameTemplateUrl: sharedIFrameConfig[shareKey].templateUrl,
+                isSharing: 'true',
```
You can try it out by simply switching it to boolean and then setting a breakpoint here: https://github.com/DanielHabenicht/jitsi-meet/blob/b7ffcae6f17cbe1bb29629f12d0633aba1b5c296/react/features/shared-iframe/middleware.any.js#L115

mihhu replied:
Oh sorry, I missed that, everything is fine then!

DanielHabenicht commented:
> > * participants that join after the iframe being shared should also see the iframe in stage view
> 
> That should already be the case, I verify that its working.

I can't reproduce this, can you send me some parameters when this happens?

mihhu commented on react/features/shared-iframe/actions.any.js:
```diff
@@ -0,0 +1,160 @@
+
+        const localParticipant = getLocalParticipant(state);
+
+        if (state['features/shared-iframe']?.iframes?.[shareKey]?.ownerId === localParticipant.id) {
```
Maybe we could extend the config to include custom permissions, WDYT? @DanielHabenicht  @saghul

DanielHabenicht replied:
Yes sounds good, it was a customer requirement so I will implement the switch.

DanielHabenicht replied:
I will add nesting layer in the config
```jsonc
sharedIFrames: {
        restrictControlToInitialSharer: true,  // <-- new option
        frames: {
          whiteboard: { // The Id of the sharedIframe
              title: 'Whiteboard',
              templateUrl: 'https://wbo.ophir.dev/boards/{room}?lang={lang}',
              avatarUrl: 'images/sharedIFrameAvatar.svg'
          },
          etherpad: {
              title: 'Document',
              templateUrl: 'https://yopad.eu/p/{room}',
              avatarUrl: 'images/sharedIFrameAvatarEtherpad.svg'
          }
       }
    },
```

mihhu commented:
> > > There are also two more things that should be covered later on:
> > > 
> > > * the more menu should be dismissed
> > 
> > 
> > I will have a look!
> 
> @mihhu Can you help me out here? I don't get why the `afterClick` event from the [OverflowMenu](https://github.com/DanielHabenicht/jitsi-meet/blob/21cf7f23c2b71eab5c36ff2f6c8c9a33fbd10953/react/features/toolbox/components/native/OverflowMenu.js#L113) is not passed down to the [SharedIFrameButtonContainer](https://github.com/DanielHabenicht/jitsi-meet/blob/b7ffcae6f17cbe1bb29629f12d0633aba1b5c296/react/features/shared-iframe/components/SharedIFrameButtonContainer.js#L31) `props`, do you happen to know why?

Oh that's because the OverflowMenu is only for the native app. I believe on the web side we don't usually manage the afterClick because it's not necessary, but in the current case, the iframe captures the click events and maybe an approach like in [this case](https://github.com/jitsi/jitsi-meet/blob/0d50f1867d3dd0ab5c4713dd50dc363259d10a50/react/features/screen-share/components/ShareAudioButton.js#L53) becomes necessary? If this doesn't work, we can table it for now and amend it later in a different PR.

mihhu commented:
> > > * participants that join after the iframe being shared should also see the iframe in stage view
> > 
> > 
> > That should already be the case, I verify that its working.
> 
> I can't reproduce this, can you send me some parameters when this happens?

Whenever I join from chrome incognito or safari in a meeting that has the whiteboard open (for example):
- the incognito and safari participants can't see the whiteboard right away (they have to click on the thumbnail)
- the firefox participant loses the whiteboard from the stage view (has to click on the thumbnail as well)

DanielHabenicht commented:
> > > > * participants that join after the iframe being shared should also see the iframe in stage view
> > > 
> > > 
> > > That should already be the case, I verify that its working.
> > 
> > 
> > I can't reproduce this, can you send me some parameters when this happens?
> 
> Whenever I join from chrome incognito or safari in a meeting that has the whiteboard open (for example):
> 
> * the incognito and safari participants can't see the whiteboard right away (they have to click on the thumbnail)
> * the firefox participant loses the whiteboard from the stage view (has to click on the thumbnail as well)

how many participants, just two or more?

mihhu commented:
> > > > > * participants that join after the iframe being shared should also see the iframe in stage view
> > > > 
> > > > 
> > > > That should already be the case, I verify that its working.
> > > 
> > > 
> > > I can't reproduce this, can you send me some parameters when this happens?
> > 
> > 
> > Whenever I join from chrome incognito or safari in a meeting that has the whiteboard open (for example):
> > 
> > * the incognito and safari participants can't see the whiteboard right away (they have to click on the thumbnail)
> > * the firefox participant loses the whiteboard from the stage view (has to click on the thumbnail as well)
> 
> how many participants, just two or more?

It got to 3-4, it seemed browser specific and the meeting started from chrome.

mihhu commented on react/features/shared-iframe/components/SharedIFrameButtonContainer.js:
```diff
@@ -0,0 +1,59 @@
+
+import SharedIFrameButton from './SharedIFrameButton';
+
+type Props = AbstractButtonProps & {
```
Please make sure to import these props.

DanielHabenicht replied:
01aaf7adc964d7bfd75f23bb02e2f1e80937c56a

DanielHabenicht commented on react/features/shared-iframe/actions.any.js:
```diff
@@ -0,0 +1,160 @@
+ *     disabled: boolean
+ * }}
+ */
+export function setDisableButton(shareKey: string, disabled: boolean) {
```
4af15f989df4ccc0af5df4956db0e6ed513da6c3

DanielHabenicht commented on react/features/shared-iframe/reducer.web.js:
```diff
@@ -0,0 +1,61 @@
+// @flow
```
I refactored the reducer, actions and middleware to use one file each, as they are the same anyway, on mobile the buttons should be disabled too.... 

4af15f989df4ccc0af5df4956db0e6ed513da6c3

DanielHabenicht commented on react/features/shared-iframe/actions.any.js:
```diff
@@ -0,0 +1,160 @@
+ * @param {string} statePatch - The state patch to be applied.
+ * @returns {Function}
+ */
+export function updateSharedIFrameState(statePatch) {
```
seems like its not, i removed it: 33424eee0bc6a916a4049d534aa629eb2bec50ee

mihhu commented:
> > > > > * participants that join after the iframe being shared should also see the iframe in stage view
> > > > 
> > > > 
> > > > That should already be the case, I verify that its working.
> > > 
> > > 
> > > I can't reproduce this, can you send me some parameters when this happens?
> > 
> > 
> > Whenever I join from chrome incognito or safari in a meeting that has the whiteboard open (for example):
> > 
> > * the incognito and safari participants can't see the whiteboard right away (they have to click on the thumbnail)
> > * the firefox participant loses the whiteboard from the stage view (has to click on the thumbnail as well)
> 
> Ahh, well that was a customer requirement too. They wanted to leave the option on what to see to the user (at least for newly joined users). Should I add another configuration option? @mihhu It would just make this line configurable: https://github.com/DanielHabenicht/jitsi-meet/blob/5d86c1cc3617565d600a62e9418fa33e02e105b5/react/features/shared-iframe/middleware.js#L172-L175
> 
> e.g.: `focusSharedIFramesOnJoin: true`

Sorry, I checked and I think I understand what's happening: the newly joined participant (e.g. the Safari one) takes the stage view. So I'm not sure here, but I would like to join a meeting without disrupting the flow for the other participants that are already engaged in co-editing something for example. You can add the option to preserve this behaviour, it can't hurt, but I'd say the current default can bother some people. WDYT, @saghul?

On another note, thanks for covering everything else! Did you get a chance to check out native as well?

saghul commented:
> Sorry, I checked and I think I understand what's happening: the newly joined participant (e.g. the Safari one) takes the stage view. So I'm not sure here, but I would like to join a meeting without disrupting the flow for the other participants that are already engaged in co-editing something for example. You can add the option to preserve this behaviour, it can't hurt, but I'd say the current default can bother some people. WDYT, @saghul?

I'd say we should behave like screen-sharing here and not change the view. It would violate the principle of least surprise for the user.

mihhu commented:
Hey @DanielHabenicht, just checking in, I believe you're already off in your sabbatical. Is it ok if I take it from here?

DanielHabenicht commented:
Yes, for sure. Sorry, my time was blocked by other priorities, so I couldn't fix the native issues.

AjayShivanagol commented:
@DanielHabenicht I have  tested this branch and I see that it has 2 issues..

1) if the config is not defined then it's giving an error. (it should work if the config is defined in config.js as well)
2) once Moderator initiates the Whiteboard other users are also able to modify the whiteboard.
the remaining participant should have only the read access.
OR access can be controlled via settings..

CC @damencho

damencho commented:
> 2\. once Moderator initiates the Whiteboard other users are also able to modify the whiteboard.
>     the remaining participant should have only the read access.
>     OR access can be controlled via settings..

That was never been on the table for discussion. There is currently a similar etherpad sharing option in jitsi-meet and if you share a document for everybody, everybody can edit it. 
Especially with the iframe, the editing is a feature of the service inside the iframe, whether it's whiteboard or some other service, so it is out of the scope of this PR.

matbgn commented:
> > 2. once Moderator initiates the Whiteboard other users are also able to modify the whiteboard.
> > the remaining participant should have only the read access.
> > OR access can be controlled via settings..
> 
> That was never been on the table for discussion. There is currently a similar etherpad sharing option in jitsi-meet and if you share a document for everybody, everybody can edit it. Especially with the iframe, the editing is a feature of the service inside the iframe, whether it's whiteboard or some other service, so it is out of the scope of this PR.

+1 please do not extend the basic implementation of this PR. KISS and functional for a first step.

As it's not clear for me what are the remaining next steps and who is in charge of what, could somebody provide a quick summary of the situation. Moreover what left to be done by whom? And as I'm in a playful mood this morning 😄 can we take the forbidden risk by talking about an hypothetic ETA for the final merge of this PR?

We are really excited to enjoy this new feature 😹

damencho commented:
@mihhu can you update, please? Wdyt?

mihhu commented:
I'll get to it next week and see what's left to be done. Afaik, I need to take a look at the native part and to fix the error @AjayShivanagol mentioned. Can't give a proper ETA yet, but the intention is to have it merged next month.

AjayShivanagol commented:
@mihhu I have tested with the latest commits changes.
1)  For the web When we open etharpad or Whiteboard we are not able to edit it.

mihhu commented:
@AjayShivanagol sorry about that, I noticed after pushing, but I was already deep in changes for the native app. 
Thanks for keeping an eye on this!

There are some things still to be figured out: 
- my native emulators aren't feeling great lately, but I think there's a re-render happening during the whiteboard edit that cuts off the tool action - can someone confirm please?
- the `restrictControlToInitialSharer` - I didn't get to it yet, but I'm not entirely sure it stops the share action from propagating, need to test
- the rebase with master, which is going to be painful at this point (with jitsi migrating to full ts and all)

I'll try to get back on this asap, any feedback is appreciated.

AjayShivanagol commented:
@mihhu I have tested this on a native app.
issue observed not being able to draw smoothly on a whiteboard. it's being cut as you mentioned.

matbgn commented:
What a huge waste of community resources 😔 So I guess this PR will never be merged? Or is there some parts that could be saved?

saghul commented:
Unfortunately we were not able to use this for the whiteboard and could not make it fit the project at this stage.
